### ISSUE TITLE
Debugger local variables frozen after Go panic

### ISSUE THREAD

**Issue by nmeocisco** [microsoft/vscode-go#2444]
I originally submitted this bug to the vscode project but was told that I need to file this bug here:

https://github.com/Microsoft/vscode/issues/72426

When debugging a Go program, if a panic occurs then the local variables drop down will be frozen with a spinning circle displayed. This stays forever so vscode has to be closed and reopened for the problem to go away.

- VSCode Version: 1.33.1
- OS Version: Ubuntu 18.04
- Go extension: ms-vscode.go 0.9.2

Steps to Reproduce:

1. Create a simple test Go file that looks like the following:

```
package main

import (
	"fmt"
	"testing"
)

func TestExample(t *testing.T) {
	s := "example"
	panic(s)
}

func TestExample2(t *testing.T) {
	fmt.Println("example")
}
```

2. Add a breakpoint on the line containing the "panic" call as shown in the following screenshot.

![image](https://user-images.githubusercontent.com/49691353/56287328-09878180-60ea-11e9-8b28-31e20793f09e.png)

3. Click on "debug test" above TestExample() to start debugging.
4. The debugger will stop at panic. Notice so far everything looks good and you can see the value of "s" shown in the local variables on the left side.

![image](https://user-images.githubusercontent.com/49691353/56287344-15734380-60ea-11e9-952e-17755aa4b83d.png)

5. Hit continue to allow the panic to occur. Notice that there is now a spinning circle next to "Local" and it is still displaying the value of "s" even though it is out of scope.

![image](https://user-images.githubusercontent.com/49691353/56287358-1e641500-60ea-11e9-8fdb-6cf914679a52.png)

6. Allow the program to continue and exit. Now start debugging TestExample2() and you will notice the local variables display is stuck. In the following screenshot it still shows a value for "s" even though TestExample2() does not have it defined.

![image](https://user-images.githubusercontent.com/49691353/56287370-26bc5000-60ea-11e9-93cd-76eeda437920.png)

The following is the response from the person who directed me to file the bug here:

"This is happening because the go extension is not returning a response for that varaibles request."

**Comment by emehrkay** [comment]
I think that I can reliably reproduce this issue

**Comment by lggomez** [comment]
> "This is happening because the go extension is not returning a response for that varaibles request."

Rather than this being solely on the debug adapter responses, shouldn't vscode clean the variables pane between sessions? because in this case it is clear that one debug session is ending anomalously for TestExample and a new one is spawned for TestExample2

These are two independent debug sessions and the debug adapter shouldn't have any state in between, even if its process hang up (I assume the main vscode process is able to kill the debug adapter process, should something this happen)

Does this repro if you reload the window before running TestExample2? (see command palette, Developer: Reload Window)

**Comment by nmeocisco** [comment]
The "Reload Window" trick appears to fix the problem. Thanks for the tip, that will save me the trouble have having to restart vscode entirely anytime this happens. Still, there is definitely a bug somewhere here that needs to be fixed since needing to manually reload the window between debug sessions shouldn't be required.

**Comment by lggomez** [comment]
Yeah, this is still at discussion at microsoft/vscode#72426 but I think this is a two stage bug:

- VSCode not refreshing/cleaning these views in-between debug sessions
- The extension not handling these requests that finish anomalously

I'm not sure on how to diagnose this from the workbench side but I will try to debug the debug adapter and see if I can improve the handling of these scenarios

CC @ramya-rao-a @jhendrixMSFT

**Comment by mingrammer** [comment]
Same here

**Comment by drew-512** [comment]
YES. please look into this VSCode heroes!   This is a regular source of pain for me.  And thank you @nmeocisco  for doing a lovely repro writeup (twice!).

As always, big thanks to the VSCode and Go teams for making amazing free product pair that just crushes it!

**Comment by lggomez** [comment]
I cannot reproduce this as of VSCode 1.36.0 in Linux, delve build 068e2451004e95d0b042e5257e34f0f08ce01466 (and version 0.11.3 of the Go extension but no relevant changes were introduced to the debug adapter in the meantime)

Can you try again after updating? @nmeocisco @mingrammer @drew-512

**Comment by nmeocisco** [comment]
Still happens for me using VSCode 1.36.0, version 0.11.3 of the Go extension, and the latest version of delve. On a panic I get the spinning circle that won't go away between debugging sessions. Tried multiple times and it always happens for me. I am on Ubuntu 18.04 if that makes a difference.

**Comment by lggomez** [comment]
It's weird that I cannot reproduce it considering that no fixes were made yet and that you still can, maybe I'm missing something here?

![Selection_026](https://user-images.githubusercontent.com/6326271/60905842-462bca80-a24c-11e9-96a3-1153f7d0414c.png)
![Selection_028](https://user-images.githubusercontent.com/6326271/60906021-a0c52680-a24c-11e9-8de6-a75066fb6539.png)
![Selection_027](https://user-images.githubusercontent.com/6326271/60905844-475cf780-a24c-11e9-8ea4-257c6042af90.png)

**Comment by nmeocisco** [comment]
No idea, but if you need any specific info from me let me know.

**Comment by lggomez** [comment]
@nmeocisco can you share the log output with the `"showLog": true` and `"trace": "verbose"` settings in your launch.json file for both debug sessions?

**Comment by nmeocisco** [comment]
Still happening with vscode 1.36.1 which I updated to this morning. Here is the output:

TestExample
```
API server listening at: 127.0.0.1:49206
2019-07-10T08:33:23-04:00 info layer=debugger launching process with args: [/home/nmeo/test/debug.test -test.run TestExample]
2019-07-10T08:33:23-04:00 info layer=debugger created breakpoint: &api.Breakpoint{ID:1, Name:"", Addr:0x511f52, File:"/home/nmeo/test/test_test.go", Line:10, FunctionName:"_/home/nmeo/test.TestExample", Cond:"", Tracepoint:false, TraceReturn:false, Goroutine:false, Stacktrace:0, Variables:[]string(nil), LoadArgs:(*api.LoadConfig)(0xc0007880c0), LoadLocals:(*api.LoadConfig)(0xc000788810), HitCount:map[string]uint64{}, TotalHitCount:0x0}
2019-07-10T08:33:23-04:00 info layer=debugger created breakpoint: &api.Breakpoint{ID:2, Name:"", Addr:0x511fe4, File:"/home/nmeo/test/test_test.go", Line:15, FunctionName:"_/home/nmeo/test.TestExample2", Cond:"", Tracepoint:false, TraceReturn:false, Goroutine:false, Stacktrace:0, Variables:[]string(nil), LoadArgs:(*api.LoadConfig)(0xc000788930), LoadLocals:(*api.LoadConfig)(0xc000788960), HitCount:map[string]uint64{}, TotalHitCount:0x0}
2019-07-10T08:33:23-04:00 debug layer=debugger continuing
2019-07-10T08:33:30-04:00 debug layer=debugger continuing
--- FAIL: TestExample (6.90s)
2019-07-10T08:33:41-04:00 debug layer=debugger continuing
panic: example [recovered]
	panic: example

goroutine 5 [running]:
testing.tRunner.func1(0xc0000ae100)
	/usr/local/go/src/testing/testing.go:792 +0x392
panic(0x52aae0, 0xc000052450)
	/usr/local/go/src/runtime/panic.go:513 +0x1e6
_/home/nmeo/test.TestExample(0xc0000ae100)
	/home/nmeo/test/test_test.go:10 +0x72
testing.tRunner(0xc0000ae100, 0x564d90)
	/usr/local/go/src/testing/testing.go:827 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x35c
2019-07-10T08:33:41-04:00 debug layer=debugger halting
```
TestExample2
```
API server listening at: 127.0.0.1:7420
2019-07-10T08:34:28-04:00 info layer=debugger launching process with args: [/home/nmeo/test/debug.test -test.run TestExample2]
2019-07-10T08:34:28-04:00 info layer=debugger created breakpoint: &api.Breakpoint{ID:1, Name:"", Addr:0x511f52, File:"/home/nmeo/test/test_test.go", Line:10, FunctionName:"_/home/nmeo/test.TestExample", Cond:"", Tracepoint:false, TraceReturn:false, Goroutine:false, Stacktrace:0, Variables:[]string(nil), LoadArgs:(*api.LoadConfig)(0xc0006f6060), LoadLocals:(*api.LoadConfig)(0xc0006f6090), HitCount:map[string]uint64{}, TotalHitCount:0x0}
2019-07-10T08:34:28-04:00 info layer=debugger created breakpoint: &api.Breakpoint{ID:2, Name:"", Addr:0x511fe4, File:"/home/nmeo/test/test_test.go", Line:15, FunctionName:"_/home/nmeo/test.TestExample2", Cond:"", Tracepoint:false, TraceReturn:false, Goroutine:false, Stacktrace:0, Variables:[]string(nil), LoadArgs:(*api.LoadConfig)(0xc0001e6300), LoadLocals:(*api.LoadConfig)(0xc0001e6330), HitCount:map[string]uint64{}, TotalHitCount:0x0}
2019-07-10T08:34:28-04:00 debug layer=debugger continuing
2019-07-10T08:34:34-04:00 debug layer=debugger continuing
example
PASS
2019-07-10T08:34:34-04:00 debug layer=debugger halting
Process exiting with code: 0
```

**Comment by drew-512** [comment]
I had it happen while working yesterday w 1.36.1 on mac OS.

**Comment by lggomezml** [comment]
This is also me on macOS, same as my Linux desktop:

![oDCZd5kipp](https://user-images.githubusercontent.com/29546751/61136530-37d2ee00-a49a-11e9-9be4-9fe9ab487193.gif)

It is a bit frustrating that I can't seem to be able to reproduce something so obvious...

**Comment by nmeocisco** [comment]
Maybe extensions make a difference? I see an icon of a whale on the left side that isn't in my environment. These are what I currently have installed as extensions:

Go 0.11.4
Git Blame 2.8.1
vscode-proto3 0.3.0

Might be worth stripping out other extensions and only using vscode with just the vscode-go extension installed.

**Comment by drew-512** [comment]
`go version go1.11.5 darwin/amd64`

![Screen Shot 2019-07-12 at 12 06 11 PM](https://user-images.githubusercontent.com/35852900/61145607-7ae29080-a49d-11e9-86dd-a03da33e337c.png)

**Comment by AdrienViala** [comment]
hello, just to point that I got the same bug too. reload windows does temporarily fix the issue. go 0.11.4 on 1.36.1 user setup.

**Comment by rracb** [comment]
Similar bug here, Windows10v1903, VSC 1.36.1 Go extension 0.11.4. Delve 0.2.0
Set breakpoint (doesn't require a panic to occur), program halts on breakpoint and the local variable spins as shown above.

No global variables but a LOT of local variables.

Doesn't happen all the time but I can't seem to pinpoint the difference between when it works and when it spins aimlessly. Will keep looking...

**Comment by nmeocisco** [comment]
@rracb that actually happens to me regularly as well but just as you say it doesn't appear to happen all the time in an easily reproducible manner. I only noticed this was consistent behavior when a panic occurred. I am hoping the fix for when this happens for the panic also fixes that behavior as well.

**Comment by wcn00** [comment]
Me too.  Really frustrating :(

**Comment by fr3fou** [comment]
any updates on this? this makes debugging painful / impossible in some cases

**Comment by ramya-rao-a** [comment]
Hello all,

I tried the exact steps that @nmeocisco has listed out in the issue description. But like @lggomez, I am not able to repro the issue. I am using

- VS Code Insiders: 1.44.0
- Go extension: 0.13.1
- go version output: `go1.12.12 darwin/amd64`
- dlv version output:
    - Version: 1.3.2
    - Build: $Id: 569ccbd514fc47c8b4c521b142556867ec5e6917 $ 

At this point, if anyone who does see this issue is interested in debugging using the source code of the extension, I'd be happy to help. Pointers:
- [Set up local dev environment](https://github.com/microsoft/vscode-go/blob/master/src/debugAdapter/Readme.md#set-up-local-dev-environment)
- [Debug the Debug adapter](https://github.com/microsoft/vscode-go/blob/master/src/debugAdapter/Readme.md#debugging-the-go-debug-adapter)
- Add breakpoint at the [variablesRequest call](https://github.com/microsoft/vscode-go/blob/master/src/debugAdapter/goDebug.ts#L1162)

The variables request call above is invoked by VS Code when you hit a breakpoint resulting in a need to update the variables pane

**Comment by ramya-rao-a** [comment]
@quoctruong Can you check if you can repro this?

**Comment by ramya-rao-a** [comment]
@quoctruong, @stamblerre, @hyangah Can any of you repro this issue?

**Comment by nmeocisco** [comment]
Just to follow up, I tried with the simple example above with vscode 1.43.1, Go extension 0.13.1, and dlv 1.3.2 and this problem no longer occurs. That said, this was originally a method I was able to make it happen in a reproducible manner 100% of the time, but I did experience it other times at random in the past when there was no panic and I can't say what method can be used to reproduce those conditions and if those scenarios might also be fixed. Maybe @fr3fou can provide more detail?
